pkg/dbmate/db.go

```diff
@@ -412,6 +415,13 @@ func (db *DB) FindMigrations() ([]Migration, error) {
 		}
 	}
 
+	var latestAppliedMigration string
+	for migrationVersion := range appliedMigrations {
+		if migrationVersion > latestAppliedMigration {
+			latestAppliedMigration = migrationVersion
+		}
```

Comment:
I will say that I think some automated tests to clarify how this change will behave could be useful to clarify expectations.

Is Go's string comparison implementation robust enough that `"100_foo" > "2_bar"` will be true?  [Looks like it isn't.](https://go.dev/play/p/0Fh_jGFUTcI)  What would you expect dbmate to do in this situation?  In this particular situation, the code above would consider "2_bar" the latest applied migration, even if "100_bar" had also been applied.  Is that what you would expect?

Reply:
The version comparison seems to be in line with how dbmate already operates. `FindMigrations()` returns a sorted slice in the end but also sorts the slice similarly just by file name. Dbmate already creates its migrations with a sortable file name, such as `20151127184807_create_users_table.sql`. So as long as all migrations are created by dbmate that will not be a problem. I agree with you that the behaviour should be properly described. The order of execution for pending migrations is described as _in numerical order_ in the [running migrations](https://github.com/amacneil/dbmate#running-migrations) section; but it's in alphanumeric order? A proper description of the flag should also be added to that section. I will look into some possible additional test cases but it seems they would also need to test the general behaviour of dmate with migration files that do not use the same file naming scheme?

Regarding the expensive loop: The applied migrations are indeed queried in descending order in all used database drivers but they are stored in an unsorted `map[string]bool` afterwards. I do not see another way to get the latest applied migration without too many unrelated changes that could affect other PRs, except doing another database query which does not seem practical.

Reply:
yeah, this looked suspicious to me as well but then I looked farther down and saw that dbmate is doing the same string comparison on filename for its sorting.

I also looked at this looping and yeah it is quite odd that the current interface returns a map of applied migrations instead of just a slice. I imagine at the time it seemed more efficient, but now it ends up being less efficient :)

Reply:
> [...] The order of execution for pending migrations is described as in numerical order in the [running migrations](https://github.com/amacneil/dbmate#running-migrations) section; but it's in alphanumeric order? [...]

Yeah, the documentation is potentially misleading, as it's in lexicographical order, which only coincidentally is in natural numeric order if the number-portion of the version number is formatted with the same number of characters, which it will be if you use dbmate-generated migration filenames, but if you want to generate your own migration filenames, this might be a source of surprise when it's not performing an actual natural sort.

Anyway, this is a potential gotcha for anyone using dbmate, not necessarily specific to this PR, so I'm happy we had this discussion but nothing actionable needs to be done with respect to this PR.  Thanks for hearing me out.

> [...] Regarding the expensive loop: The applied migrations are indeed queried in descending order in all used database drivers but they are stored in an unsorted map[string]bool afterwards. [...]

You are right, my mistake.

It's unfortunate that Go doesn't ship with an ordered map implementation out of the box.

I suppose having `Driver.SelectMigrations()` return a `[]string` slice is unappealing since using `sort.Search()` to test for existence of elements in the slice is a bit clumsy compared to indexing into a map and checking the return value, and having an O(1) operation to check for existence is probably more important since that's probably the more frequently executed use case, anyway.